README.md
# rscript <img src="man/figure/rscript.png" align="right" width="120" />
![lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-lightgrey)![status:
development](https://img.shields.io/badge/status-development-blue)

# Overview

Create and comment in script R. The creation of the r script inserts the information from the code development environment such as code editor, platform, author name, among others. Comments can be called interactively with output on the clipboard or in console.

# Installation

- CRAN

You can install the released version of rscript from [CRAN](https://CRAN.R-project.org) with:

``` r
# Wait.
```

- Github

Or install the development version of [Github](https://github.com/):

```
remotes::install_github("brunofariadf/rscript")
```

# Usage

This is a basic example that shows how to use a package function.

- R
``` r
# count line of code in the script
rscript::script_count(list.files("./R", full.names = TRUE))[1:4]
#    check.R engineer.R    error.R     head.R
#        143         89         60         77
```

``` r
# create header to script
rscript::script_head(clipboard = FALSE)
#  [1] "## ------------------------------------------"
#  [2] "##                          Date: 2022-07-28"
#  [3] "## ------------------------------------------"
#  [4] "## Project: "
#  [5] "## Author: "
#  [6] "## "
#  [7] "## GUI: VSCode (1.69.2)"
#  [8] "## Language: R (4.1.0)"
#  [9] "## Platform: Windows"
# [10] "## ------------------------------------------"
```

``` r
# create header to script with python or julia
rscript::script_head(type_language = "python", clipboard = FALSE)
#  [1] "## ------------------------------------------"
#  [2] "##                          Date: 2022-07-28"
#  [3] "## ------------------------------------------"
#  [4] "## Project: "
#  [5] "## Author: "
#  [6] "## "
#  [7] "## GUI: VSCode (1.69.2)"
#  [8] "## Language: Python (3.9.6)"
#  [9] "## Platform: Windows"
# [10] "## ------------------------------------------"

rscript::script_head(type_language = "julia", clipboard = FALSE)
#  [1] "## ------------------------------------------"
#  [2] "##                          Date: 2022-07-28"
#  [3] "## ------------------------------------------"
#  [4] "## Project: "
#  [5] "## Author: "
#  [6] "## "
#  [7] "## GUI: VSCode (1.69.2)"
#  [8] "## Language: Julia (1.7.3)"
#  [9] "## Platform: Windows"
# [10] "## ------------------------------------------"
```

```r
# create 10 R script
script_create("./tests", name_script = letters[1:10])
fs::dir_tree("./tests")
# ./tests
# +-- a.R
# +-- b.R
# +-- c.R
# +-- d.R
# +-- e.R
# +-- f.R
# +-- g.R
# +-- h.R
# +-- i.R
# \-- j.R
```

# Platform

Development.

<div align="center">

|platform | available |
|:--------|:---------:|
|Windows  |    yes    |
|Linux    |    no     |
|MacOS    |    no     |

</div>

# Code of Conduct

Please note that the project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html). By contributing to this project, you agree to abide by its terms.

# License

MIT

NEWS.md
# rscript 0.0.0.9000 (development)

## Note from version

### &gt; rscript (development)

-   development

### &gt; rscript 0.0.0.9000 (2022-08-01)

-   Start of the package.


R/is.R
## ------------------------------------------
##                          Date: 2022-07-26
## ------------------------------------------
## Project: rscript
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Language: R (4.1.0)
## Platform: windows (10.0.14393)
## ------------------------------------------

## search the platform
## not export

.is_windows <- function() {
    w_i <- .Platform[["OS.type"]]
    if (w_i == "windows") TRUE else FALSE
}

.is_unix <- function() {
    x_i <- .Platform[["OS.type"]]
    if (x_i == "unix") TRUE else FALSE
}

# .is_reditorsupport <- function() {
#     if (.is_windows()) {
#         editor_i <- Sys.getenv()[["VSCODE_INIT_R"]]
#         grepl(editor_i, pattern = "reditorsupport")
#     }
# }

.is_vscode <- function() {
    if (.is_windows()) {
        if (.exists_vscode()) {
            check_i <- search()
            is.element("tools:vscode", check_i)
        }
    }
}

.is_rstudio <- function() {
    if (.is_windows()) {
        if (.exists_rstudio()) {
            check_i <- search()
            is.element("tools:rstudio", check_i)
        }
    }
}

## search the software
## not export

.exists_vscode <- function() {
    if (.is_windows()) {
        x_i <- shell("where code", intern = TRUE)
        x_i <- file.exists(x_i)
        is.element(TRUE, x_i)
    }
}

.exists_rstudio <- function() {
    if (.is_windows()) {
        program_i <- dirname(Sys.getenv("CommonProgramFiles"))
        path_i <- file.path(program_i, "RStudio", "bin", "rstudio.exe")
        path_i <- suppressWarnings(normalizePath(path_i))
        file.exists(path_i)
    }
}

# .exists_rstudio()

.exists_python <- function() {
    if (.is_windows()) {
        p_i <- shell("where python", intern = TRUE)
        is.element("python.exe", basename(p_i))
    }
}

.exists_julia <- function() {
    if (.is_windows()) {
        j_i <- shell("where julia", intern = TRUE)
        is.element("julia.exe", basename(j_i))
    }
}

.exists_xclip <- function() {
    if (.is_unix()) {
        x_i <- system("which xclip", intern = TRUE)
        # x_i <- file.exists(system("which xclip", intern = TRUE))
        x_i <- file.exists(Sys.which("xclip")[1L])
        if (isTRUE(x_i)) TRUE else FALSE
    }
}

# extract extension in the file
# not export

.exists_extension <- function(path) {
    ext_i <- tolower(tools::file_ext(path))
    ext_i <- is.element("r", ext_i)
    all(ext_i)
}


R/main.R
## ------------------------------------------
##                          Date: 2022-07-26
## ------------------------------------------
## Project: rscript
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Language: R (4.1.0)
## Platform: windows (10.0.14393)
## ------------------------------------------

#' Header to R Script
#'
#' Creates header with information from the code development environment, author, and more.
#'
#' @usage script_head(name_author = NULL, 
#'  name_project = NULL, 
#'  type_language = "r", 
#'  type_line = "-", 
#'  type_size = 42, 
#'  add_line = NULL, 
#'  rm_line = NULL, 
#'  clipboard = TRUE)
#' @param name_author character class vector that indicates the author name.
#' @param name_project character class vector that indicates the author project.
#' @param type_language character class vector that indicates the language version with option 'r', 'python' and 'julia'.
#' @param type_line character class vector that indicates the format type to header.
#' @param type_size numeric class vector that indicates the length to header.
#' @param add_line numeric class vector that indicates the vector to add in the header.
#' @param rm_line numeric class vector that indicates the vector to remove in the header.
#' @param clipboard logical class vector that indicates output to clipboard or console with option 'TRUE' or 'FALSE'.
#' @return character class vector with code development environment information.
#' @export 
#' @examples
#' # default
#' # script_head()
#' 
#' # usage other
#' script_head(clipboard = FALSE)
#' script_head(name_author = "test", clipboard = FALSE)
#' script_head(name_author = c("test 1", "test 2", "test 3"), clipboard = FALSE)
#' script_head(name_project = "test 2", clipboard = FALSE)
#' script_head(add_line = "Description: test", 
#'  rm_line = c("Platform", "GUI", "Author"), 
#'  clipboard = FALSE)

script_head <- function(name_author = NULL, name_project = NULL, 
    type_language = "r", type_line = "-", type_size = 42, 
    add_line = NULL, rm_line = NULL, clipboard = TRUE) {

    .check_script_head(name_author, name_project, 
        type_language, type_line, type_size, add_line, rm_line, clipboard)
}

#' Comment Type to Script R
#'
#' Creates title and subtitle as comment to organize the R script.
#'
#' @usage script_comment(x, 
#'  type_comment = 'h1',
#'  type_line = '-',
#'  type_size = NULL,
#'  clipboard = TRUE)
#' @param x character class vector that indicates the message of the title or subtitle of the comment.
#' @param type_comment character class vector that indicates the comment type with option 'h1', 'h2' or 'h3'.
#' @param type_line character class vector that indicates the format type to header.
#' @param type_size numeric class vector that indicates the length to header.
#' @param clipboard logical class vector that indicates output to clipboard or console with option 'TRUE' or 'FALSE'.
#' @return character class vector that indicates the comment type for R script.
#' @export 
#' @examples
#' # default
#' # script_comment("test")
#' 
#' # other usage
#' # h1
#' script_comment('test_1', clipboard = FALSE)
#' script_comment(c("test_1", "test_2"), clipboard = FALSE)
#' 
#' # h2
#' script_comment('test_1', type_comment = "h2", clipboard = FALSE)
#' script_comment(c("test_1", "test_2"), type_comment = "h2", clipboard = FALSE)
#' 
#' # h3
#' script_comment('test_1', type_comment = "h3", clipboard = FALSE)
#' script_comment(c("test_1", "test_2"), 
#'  type_comment = "h3", type_size = 40, clipboard = FALSE)

script_comment <- function(x, type_comment = "h1", 
    type_line = "-", type_size = NULL, clipboard = TRUE) {
    .check_script_comment(x, type_comment, type_line, type_size, clipboard)
}

#' Create R Script
#'
#' Creates the R script with the information header of the code development environment, author, and more.
#'
#' @usage script_create(path, 
#'  name_script, 
#'  name_author = NULL, 
#'  name_project = NULL, 
#'  type_line = "-", 
#'  type_size = 42, 
#'  add_line = NULL, 
#'  rm_line = NULL)
#' @param path character class vector that indicates the directory path.
#' @param name_script character class vector that indicates the file name.
#' @param name_author character class vector that indicates the author name.
#' @param name_project character class vector that indicates the author project.
#' @param type_line character class vector that indicates the format type to header.
#' @param type_size numeric class vector that indicates the length to header.
#' @param add_line numeric class vector that indicates the vector to add in the header.
#' @param rm_line numeric class vector that indicates the vector to remove in the header.
#' @return the R script file with code development environment information in header.
#' @export 
#' @examples
#' # script_create(getwd(), "test 1")
#' # script_create(getwd(), c("test 1", "test 2"))

script_create <- function(path, name_script, name_author = NULL, name_project = NULL, 
        type_line = "-", type_size = 42, add_line = NULL, rm_line = NULL) {

    if (dir.exists(path)) {
        invisible(lapply(name_script, function(i).check_script_create(path, i, 
            name_author, name_project, type_language = "r", 
            type_line, type_size, add_line, rm_line)))
    } else {
        stop(.error("directory", path))
    }
}

#' Count Line of Code
#'
#' Count of code line written in script R. The default counting process ignores comment and empty.
#'
#' @usage script_count(path, exclude = "#|^$")
#' @param path character class vector that indicates the file path.
#' @param exclude character class vector that indicates the default to exclude with regular expression.
#' @return character class vector that indicates the code line count for R script.
#' @export 
#' @examples
#' # default
#' # script_count(path)

script_count <- function(path, exclude = "#|^$") {
    unlist(lapply(path, function(x).check_script_count(x, exclude)))
}


R/check.R
## ------------------------------------------
##                          Date: 2022-07-26
## ------------------------------------------
## Project: rscript
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Language: R (4.1.0)
## Platform: windows (10.0.14393)
## ------------------------------------------

# check head
# not export

.check_script_head <- function(name_author, name_project, 
    type_language, type_line, type_size, add_line, rm_line, clipboard) {

    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    if (!is.null(name_author)) {
        if (!is.character(name_author)) {
            stop(.error("character", name_author))
        }
    }

    if (!is.null(name_project)) {
        if (!is.character(name_project)) {
            stop(.error("character", name_project))
        }
    }

    if (!is.character(type_language)) {
        stop(.error("character", type_language))
    }

    if (!is.element(tolower(type_language), c("r", "python", "julia"))) {
        stop(.error("language", type_language))
    }

    if (tolower(type_language) == "python") {
        if (!.exists_python()) {
            stop(.error("python", type_language))
        }
    }

    if (tolower(type_language) == "julia") {
        if (!.exists_julia()) {
            stop(.error("julia", type_language))
        }
    }

    if (!is.character(type_line)) {
        stop(.error("character", type_line))
    }

    if (!nchar(type_line) == 1 && length(type_line) == 1) {
        stop(.error("n_char", type_line))
    }

    if (!is.numeric(type_size)) {
        stop(.error("numeric", type_size))
    }

    if (!is.null(add_line)) {
        if (!is.character(add_line)) {
            stop(.error("character", add_line))
        }
    }

    if (!is.null(rm_line)) {
        if (!is.character(rm_line)) {
            stop(.error("character", rm_line))
        }
    }

    if (!is.logical(clipboard)) {
        stop(.error("logical", clipboard))
    }

    if (!is.element(clipboard, c(TRUE, FALSE))) {
        stop(.error("true_false", clipboard))
    }

    .get_script_head(name_author, name_project, 
        type_language, type_line, type_size, add_line, rm_line, clipboard)
}

# check comment
# not export

.check_script_comment <- function(x, type_comment, type_line, type_size, clipboard) {

    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    if (!is.character(x)) {
        stop(.error("character", x))
    }

    if (!is.character(type_comment)) {
        stop(.error("character", type_comment))
    }

    if (!is.element(type_comment, c("h1", "h2", "h3"))) {
        stop(.error("comment", type_comment))
    }

    if (!is.character(type_line)) {
        stop(.error("character", type_line))
    }

    if (!nchar(type_line) == 1 && length(type_line) == 1) {
        stop(.error("n_char", type_line))
    }

    if (!is.null(type_size)) {
        if (!is.numeric(type_size)) {
            stop(.error("numeric", type_size))
        }
    }

    if (!is.logical(clipboard)) {
        stop(.error("logical", clipboard))
    }

    if (!is.element(clipboard, c(TRUE, FALSE))) {
        stop(.error("true_false", clipboard))
    }

    .get_check_script_comment(x, type_comment, type_line, type_size, clipboard)
}

# check create
# not export

.check_script_create <- function(path, name_script, name_author, name_project, 
        type_language, type_line, type_size, add_line, rm_line) {

    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    if (!is.character(name_script)) {
        stop(.error("character", name_script))
    }

    if (!is.null(name_author)) {
        if (!is.character(name_author)) {
            stop(.error("character", name_author))
        }
    }

    if (!is.null(name_project)) {
        if (!is.character(name_project)) {
            stop(.error("character", name_project))
        }
    }

    if (!is.character(type_language)) {
        stop(.error("character", type_language))
    }

    if (!is.character(type_line)) {
        stop(.error("character", type_line))
    }

    if (!nchar(type_line) == 1 && length(type_line) == 1) {
        stop(.error("n_char", type_line))
    }

    if (!is.numeric(type_size)) {
        stop(.error("numeric", type_size))
    }

    if (!is.null(add_line)) {
        if (!is.character(add_line)) {
            stop(.error("character", add_line))
        }
    }

    if (!is.null(rm_line)) {
        if (!is.character(rm_line)) {
            stop(.error("character", rm_line))
        }
    }

    .get_script_create(path, name_script, name_author, name_project, 
        type_language, type_line, type_size, add_line, rm_line)
}

# check count
# not export

.check_script_count <- function(path, exclude) {

    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    if (!is.character(path)) {
        stop(.error("character", path))
    }

    if (!file.exists(path)) {
        stop(.error("file_exists", path))
    }

    if (!.exists_extension(path)) {
        path_i <- toString(basename(path))
        stop(.error("extension", path_i))
    }

    if (!is.character(exclude)) {
        stop(.error("character", exclude))
    }

    .get_script_count(path, exclude)
}


R/error.R
## ------------------------------------------
##                          Date: 2022-07-27
## ------------------------------------------
## Project: rscript
## Author: Bruno Faria
## 
## GUI: VSCode (1.69.2)
## Language: R (4.1.0)
## Platform: Windows
## ------------------------------------------

# check error
# not export

.error_platform <- function(insert) {
    p_i <- .Platform[["OS.type"]]
    paste0("Method not applied to the ", p_i, " platform.")
}

.error_directory <- function(insert) {
    paste0("'", insert, "' must be character class vector and the a existing directory.")
}

.error_character <- function(insert) {
    paste0("'", insert, "' must be character class vector.")
}

.error_logical <- function(insert) {
    paste0("'", insert, "' must be logical class vector.")
}

.error_numeric <- function(insert) {
    paste0("'", insert, "' must be numeric class vector.")
}

.error_language <- function(insert) {
    paste0("'", insert, "' must be character class vector with option 'r', 'python' or 'julia'.")
}

.error_n_char <- function(insert) {
    paste0("'", insert, "' must be character class vector to length and letter equal 1.")
}

.error_true_false <- function(insert) {
    paste0("'", insert, "' must be logical class vector with option 'TRUE' or 'FALSE'.")
}

.error_comment <- function(insert) {
    paste0("'", insert, "' must be character class vector with option 'h1', 'h2' or 'h3'.")        
}

.error_file_exists <- function(insert) {
    paste0("'", insert, "' must be character class vector and an existing file.")
}

.error_extension <- function(insert) {
    paste0("'", insert, "' must be character class vector with extension file 'r' or 'R'.")
}

.error_python <- function(insert) {
    paste0("Could not find the installation of the '", insert, "'")
}

.error_julia <- function(insert) {
    paste0("Could not find the installation of the '", insert, "'")
}

.error <- function(type, insert) {
    switch(type,
        "platform" = .error_platform(insert),
        "directory" = .error_directory(insert),
        "character" = .error_character(insert),
        "logical" = .error_logical(insert),
        "numeric" = .error_numeric(insert),
        "language" = .error_language(insert),
        "n_char" = .error_n_char(insert),
        "true_false" = .error_true_false(insert),
        "comment" = .error_comment(insert),
        "file_exists" = .error_file_exists(insert),
        "extension" = .error_extension(insert),
        "python" = .error_python(insert),
        "julia" = .error_julia(insert),
        "Unknown error type."
    )
}


R/utils.R
## ------------------------------------------
##                          Date: 2022-07-26
## ------------------------------------------
## Project: rscript
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Language: R (4.1.0)
## Platform: windows (10.0.14393)
## ------------------------------------------

# format gui
# not export

.format_gui <- function() {
    tryCatch(.get_gui(), error = function(i)"")
}

#.format_gui()

.format_language <- function(type) {
    tryCatch(.get_language(type), error = function(i)"")
}

# .format_language("julia")

## format the  platform with the version to create the script
## not export

.format_platform <- function() {
    if (.is_windows()) {
        as.vector(Sys.info()["sysname"])
    } else if (.is_unix()) {
        as.vector(Sys.info()["sysname"])
    }
}

# .format_platform()

# export clipboard
# not export

.write_clip <- function(x, clipboard) {
    if (isTRUE(clipboard)) {
        clipr::write_clip(x)
        message("Copied.")
    } else if (isFALSE(clipboard)) {
        return(x)
    }
}

## export r
## not export

.write_r <- function(x, path, name_script) {
    path_i <- file.path(path, paste0(name_script, ".R"))
    path_i <- suppressWarnings(normalizePath(path_i))
    writeLines(x, con = path_i, sep = "\n")
}

# check version
# not export

.count_version <- function(path, date, dev = TRUE) {
    .version <- function(x) {
        if (unique(x) == 0) {
            return(0)
        } else {
            return(max(x))
        }
    }

    if (file.exists(path)) {
        if (date == "start") {
            return("0.0.0.9000")
        } else {
            df_i <- utils::read.csv2(path, sep = ",")
            df_i <- if (date == "all") df_i else df_i[df_i[["date"]] <= as.Date(date),]
            major_i <- .version(df_i[["version"]])
            minor_i <- sum(grepl(df_i[["type"]], pattern = "minor"))
            patch_i <- sum(grepl(df_i[["type"]], pattern = "patch"))
            dev_i <- if (isTRUE(dev)) ".9000" else ""
            paste0(major_i, ".", minor_i, ".", patch_i, dev_i, " (", date, ")")
        }
    } else {
        stop("unknown file path.")
    }
}


R/engineer.R
## ------------------------------------------
##                          Date: 2022-07-26
## ------------------------------------------
## Project: rscript
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Language: R (4.1.0)
## Platform: windows (10.0.14393)
## ------------------------------------------

## ------------------------------------------
## engineer
## ------------------------------------------

# get comment
# not export

.get_check_script_comment <- function(x, type_comment, type_line, type_size, clipboard) {
    if (type_comment == "h1") {
        if (is.null(type_size)) {
            comment_i <- unlist(lapply(x, 
                function(i).join_comment_format(i, type_comment, type_line, type_size = 80)))
            .write_clip(comment_i, clipboard)
        } else if (!is.null(type_size)) {
            comment_i <- unlist(lapply(x, 
                function(i).join_comment_format(i, type_comment, type_line, type_size)))
            .write_clip(comment_i, clipboard)
        }
    } else if (type_comment == "h2") {
        if (is.null(type_size)) {
            comment_i <- unlist(lapply(x, 
                function(i).join_comment_format(i, type_comment, type_line, type_size = 42)))
            .write_clip(comment_i, clipboard)
        } else if (!is.null(type_size)) {
            comment_i <- unlist(lapply(x, 
                function(i).join_comment_format(i, type_comment, type_line, type_size)))
            .write_clip(comment_i, clipboard)
        }
    } else if (type_comment == "h3") {
        if (is.null(type_size)) {
            comment_i <- unlist(lapply(x, 
                function(i).join_comment_format(i, type_comment, type_line, type_size = 3)))
            .write_clip(comment_i, clipboard)
        } else if (!is.null(type_size)) {
            comment_i <- unlist(lapply(x, 
                function(i).join_comment_format(i, type_comment, type_line, type_size)))
            .write_clip(comment_i, clipboard)
        }
    }
}

# .get_check_script_comment(c("1", "2"), "h1", "-", NULL, TRUE)
# .get_check_script_comment(c("1", "2"), "h2", "-", NULL, TRUE)
# .get_check_script_comment(c("1", "2"), "h3", "-", NULL, TRUE)

# get create
# not export

.get_script_create <- function(path, name_script, name_author, name_project, 
            type_language, type_line, type_size, 
            add_line, rm_line) {
    .write_r(
        .get_script_head(name_author, name_project, 
            type_language, type_line, type_size, 
            add_line, rm_line, clipboard = FALSE),
        path, name_script
    )
}

# .join_comment_format("test", "h1", "-", 80)
# .join_comment_format("test", "h2", "-", 40)
# .join_comment_format("test", "h3", "-", 3)

# count line
# not export

.get_script_count <- function(path, exclude) {
    file_i <- trimws(suppressWarnings(readLines(path)))
    file_i <- Filter(function(x)!grepl(x, pattern = exclude), file_i)
    n_i = length(file_i)
    names(n_i) <- basename(path)
    return(n_i)
}

## ------------------------------------------
## utils engineer
## ------------------------------------------

## format comment to script
## not export

.join_comment_format <- function(x, type_comment, type_line, type_size) {
    if (is.element(type_comment, c("h1", "h2"))) {
        c(
            .format_comment_structure(type_line, type_size),
            .format_content_comment(x, type_comment),
            .format_comment_structure(type_line, type_size)
        )
    } else if (type_comment == "h3") {
        h3_i <- paste0(rep(type_line, type_size), collapse = "")
        .format_comment_line(paste0(x, " ", h3_i))
    }
}

# get gui
# not export

.get_gui <- function() {
    if (.is_windows()) {
        if (.is_vscode()) {
            .vsc_ver <- shell("code --version", intern = TRUE)[1]
            paste0("VSCode", " (", .vsc_ver, ")")
        } else if (.is_rstudio()) {
            program_i <- dirname(Sys.getenv("CommonProgramFiles"))
            path_i <- file.path(program_i, "RStudio", "VERSION")
            path_i <- suppressWarnings(normalizePath(path_i))
            .rstudio_ver <- suppressWarnings(readLines(path_i))
            paste0("RStudio", " (", .rstudio_ver, ")")
        } else {
            .Platform[["GUI"]]
        }
    }
}

# get language
# not export

.get_language <- function(type) {
    if (.is_windows()) {
        if (tolower(type) == "r") {
            r_i <- paste0(R.Version()[["major"]], ".", R.Version()[["minor"]])
            paste0("R (", r_i, ")")
        } else if (tolower(type) == "python") {
            p_i <- shell("python --version", intern = TRUE)
            paste0(gsub("\\ ", "\\ (", p_i), ")")
        } else if (tolower(type) == "julia") {
            j_i <- shell("julia --version", intern = TRUE)
            j_i <- paste0(gsub("\\bversion\\b ", "\\(", j_i), ")")
            gsub("\\bjulia\\b", "Julia", j_i)
        }
    }
}


inst/scripts/build.R
## ------------------------------------------
##                          Date: 2022-07-28
## ------------------------------------------
## Project: vscoder
## Author: Bruno Faria
## 
## GUI: VSCode (1.69.2)
## Language: R (4.1.0)
## Platform: Windows
## ------------------------------------------

## --------------------------------------------------------------------------------
## set variable - date: 2022-07-28 15:44:11
## --------------------------------------------------------------------------------

path_csv <- file.path("inst", "exdata", "version", "news.csv")
path_rmd <- file.path("inst", "exdata", "version", "news.Rmd")
path_md <- file.path("inst", "exdata", "version", "NEWS.md")
path_news <- file.path(getwd(), "NEWS.md")

## --------------------------------------------------------------------------------
## News - date: 2022-07-28 15:47:39
## --------------------------------------------------------------------------------

rmarkdown::render(input = path_rmd, output_file = path_news)

## ------------------------------------------
## remove
## ------------------------------------------

rm(path_csv, path_md, path_news, path_rmd)


R/head.R
## ------------------------------------------
##                          Date: 2022-07-26
## ------------------------------------------
## Project: rscript
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Language: R (4.1.0)
## Platform: windows (10.0.14393)
## ------------------------------------------

## format comment to script
## not export

.format_comment_line <- function(...) {
    paste0("## ", ...)
}

# .format_comment_line(c("test", "test 2"))

## format the structure from comment to script
## not export

.format_comment_structure <- function(type_line, type_size) {
    .format_comment_line(paste0(rep(type_line, type_size), collapse = ""))
}

# .format_comment_structure("-", 80)

## format comment to script
## not export

.format_content_comment <- function(x, type_comment) {
    if (type_comment == "h1") {
        paste0(.format_comment_line(x), " - date: ", format.POSIXct(Sys.time()))
    } else if (type_comment == "h2") {
        paste0(.format_comment_line(x))
    } else if (type_comment == "h3") {
        x
    }
}

# .format_content_comment("test", "h1")
# .format_content_comment("test", "h2")
# .format_content_comment("test", "h3")

## format comment to script create
## not export

.format_content_create_head <- function() {
    .format_comment_line(gsub("\\,", "", toString(rep(" ", 13))), "Date: ", Sys.Date())
}

# .format_content_create_head()

## format comment to script create
## not export

.format_content_create_basic <- function(name_author, name_project, type_language, add_line) {
    .format_comment_line(
        c(
            paste0("Project: ", name_project),
            paste0("Author: ", name_author),
            paste0(""),
            paste0("GUI: ", .format_gui()),
            paste0("Language: ", .format_language(type_language)),
            paste0("Platform: ", .format_platform()),
            add_line
        )
    )
}

# .format_content_create_basic("Bruno", "test", "julia", NULL)

## format content to script
## not export

.join_content_format <- function(name_author, name_project, type_language, 
    type_line, type_size, add_line) {
    c(
        .format_comment_structure(type_line, type_size),
        .format_content_create_head(),
        .format_comment_structure(type_line, type_size),
        .format_content_create_basic(name_author, name_project, type_language, add_line),
        .format_comment_structure(type_line, type_size)
    )
}

# .join_content_format("Bruno", "Test", "julia", "-", 42, NULL)

.join_content_format_rm <- function(name_author, name_project, 
    type_language, type_line, type_size, add_line, rm_line) {
    basic_i <- .join_content_format(name_author, name_project, type_language, 
        type_line, type_size, add_line)
    Filter(function(x)!grepl(x, pattern = paste0(rm_line, collapse = "|")), basic_i)
}

# .join_content_format_rm("Bruno", "Test", "julia", "-", 42, NULL, "Platform")

.join_content_format_add <- function(name_author, name_project, 
    type_language, type_line, type_size, add_line) {
    .join_content_format(name_author, name_project, 
        type_language, type_line, type_size, add_line)
}

# .join_content_format_add("Bruno", "Test", "julia", "-", 42, "Welcome!")

.join_content_format_rm_add <- function(name_author, name_project, 
    type_language, type_line, type_size, add_line, rm_line) {
    add_i <- .join_content_format_add(name_author, name_project, 
        type_language, type_line, type_size, add_line)
    Filter(function(x)!grepl(x, pattern = paste0(rm_line, collapse = "|")), add_i)
}

# .join_content_format_rm_add("Bruno", "Test", "julia", "-", 42, "Welcome!", c("Platform", "GUI"))

# engineer head
## not export

.get_script_head <- function(name_author, name_project, 
    type_language, type_line, type_size, add_line, rm_line, clipboard) {
    if (is.null(add_line) && is.null(rm_line)) {
        h_i <- .join_content_format(name_author, name_project, 
            type_language, type_line, type_size, add_line)
        .write_clip(h_i, clipboard)
    } else if (!is.null(add_line) && is.null(rm_line)) {
        h_i <- .join_content_format_add(name_author, name_project, 
            type_language, type_line, type_size, add_line)
        .write_clip(h_i, clipboard)
    } else if (is.null(add_line) && !is.null(rm_line)) {
        h_i <- .join_content_format_rm(name_author, name_project, 
            type_language, type_line, type_size, add_line, rm_line)
        .write_clip(h_i, clipboard)
    } else if (!is.null(add_line) && !is.null(rm_line)) {
        h_i <- .join_content_format_rm_add(name_author, name_project, 
            type_language, type_line, type_size, add_line, rm_line)
        .write_clip(h_i, clipboard)
    }
}
